### PR TITLE
chore: Prepare 42.0.0 Release

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

Part of https://github.com/apache/datafusion/issues/11902

## What changes are included in this PR?

- Bump version
- [Changelog](https://github.com/andygrove/datafusion/blob/prep-42/dev/changelog/42.0.0.md)

### CODE REVIEW

File: dev/changelog/42.0.0.md [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+- Minor: Add a test for version() function [#12441](https://github.com/apache/datafusion/pull/12441) (alamb)
+- Handle case-sensitive identifier when decorrelating predicate subquery [#12443](https://github.com/apache/datafusion/pull/12443) (goldmedal)
+- Bump send and express in /datafusion/wasmtest/datafusion-wasm-app [#12447](https://github.com/apache/datafusion/pull/12447) (dependabot[bot])
+- Add PartialOrd for the DF subfields/structs for the WindowFunction expr [#12421](https://github.com/apache/datafusion/pull/12421) (ngli-me)
+- Making avro_to_arrow::schema::to_arrow_schema public [#12452](https://github.com/apache/datafusion/pull/12452) (ameyc)
+- Bump rustyline from 11.0.0 to 14.0.0 in /datafusion-cli [#12407](https://github.com/apache/datafusion/pull/12407) (dependabot[bot])
+- Minor: add `ListingOptions::with_file_extension_opt` [#12461](https://github.com/apache/datafusion/pull/12461) (alamb)
+- Improve PhysicalExpr and Column documentation [#12457](https://github.com/apache/datafusion/pull/12457) (alamb)
+- fix length error with `array_has` [#12459](https://github.com/apache/datafusion/pull/12459) (samuelcolvin)
```
Comment (samuelcolvin) [inline dev/changelog/42.0.0.md:336]: this is a bug fix.

Comment (andygrove) [reply dev/changelog/42.0.0.md:336]: The changelog script categorizes PRs based on GitHub labels and/or [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/) prefixes, so PRs with the `bug` label or a commit message starting with `fix:` would get categorized as bug fixes.

Comment (andygrove) [reply dev/changelog/42.0.0.md:336]: Script for reference: https://github.com/apache/datafusion/blob/main/dev/release/generate-changelog.py

Comment (andygrove) [reply dev/changelog/42.0.0.md:336]: @samuelcolvin  Let me know if you want to add labels to any PRs and I can regenerate the changelog in this PR

Comment (samuelcolvin) [reply dev/changelog/42.0.0.md:336]: Ok, no worries, I don't have permissions to edit labels.

No need to change it for me.

Thanks as ever for datafusion.

Comment (andygrove) [reply dev/changelog/42.0.0.md:336]: I used to review all the PRs when generating changelogs, but with ~300 per release, it got a bit much. It looks like we haven't documented how the changelog generation works in the contributor guide. I will add that to my TODO list.

**Review by comphead — APPROVED** [review]
lgtm thanks @andygrove

**Pull request merged**